datafusion/core/tests/sqllogictests/src/main.rs

```diff
@@ -15,144 +15,162 @@
+use crate::engines::postgres;
+use crate::engines::postgres::image::{PG_DB, PG_PASSWORD, PG_PORT, PG_USER};
+use crate::engines::postgres::Postgres;
 
-    /// Engine name of current database.
-    fn engine_name(&self) -> &str {
-        "DataFusion"
-    }
+mod engines;
+mod setup;
+mod utils;
 
-    /// [`Runner`] calls this function to perform sleep.
-    ///
-    /// The default implementation is `std::thread::sleep`, which is universial to any async runtime
-    /// but would block the current thread. If you are running in tokio runtime, you should override
-    /// this by `tokio::time::sleep`.
-    async fn sleep(dur: Duration) {
-        tokio::time::sleep(dur).await;
-    }
-}
+const TEST_DIRECTORY: &str = "tests/sqllogictests/test_files";
+const PG_COMPAT_FILE_PREFIX: &str = "pg_compat_";
 
 #[tokio::main]
 #[cfg(target_family = "windows")]
-pub async fn main() -> Result<()> {
+pub async fn main() -> Result<(), Box<dyn Error>> {
     println!("Skipping test on windows");
     Ok(())
 }
 
 #[tokio::main]
 #[cfg(not(target_family = "windows"))]
-pub async fn main() -> Result<()> {
+pub async fn main() -> Result<(), Box<dyn Error>> {
     // Enable logging (e.g. set RUST_LOG=debug to see debug logs)
-
-    use sqllogictest::{default_validator, update_test_file};
     env_logger::init();
 
     let options = Options::new();
 
-    // default to all files in test directory filtering based on name
-    let files: Vec<_> = std::fs::read_dir(TEST_DIRECTORY)
-        .unwrap()
-        .map(|path| path.unwrap().path())
-        .filter(|path| options.check_test_file(path.as_path()))
-        .collect();
+    let files: Vec<_> = read_test_files(&options);
 
     info!("Running test files {:?}", files);
 
     for path in files {
-        println!("Running: {}", path.display());
-
         let file_name = path.file_name().unwrap().to_str().unwrap().to_string();
+        let is_pg_compatibility_test = file_name.starts_with(PG_COMPAT_FILE_PREFIX);
 
-        // Create the test runner
-        let ctx = context_for_test_file(&file_name).await;
-        let mut runner = sqllogictest::Runner::new(DataFusion { ctx, file_name });
-
-        // run each file using its own new DB
-        //
-        // We could run these tests in parallel eventually if we wanted.
         if options.complete_mode {
-            info!("Using complete mode to complete {}", path.display());
-            let col_separator = " ";
-            let validator = default_validator;
-            update_test_file(path, runner, col_separator, validator)
-                .await
-                .map_err(|e| e.to_string())?;
+            run_complete_file(&path, file_name, is_pg_compatibility_test).await?;
+        } else if options.postgres_runner {
+            if is_pg_compatibility_test {
+                run_test_file_with_postgres(&path, file_name).await?;
+            } else {
+                debug!("Skipping test file {:?}", path);
+            }
```

Comment:
The tests which name starts with `pg_compat_` would run with Datafusion during `cargo test`.
In order to check compatibility with Postgres, there is a github job that runs these tests with a Postgres runner.

Thus Datafusion runs all the time. And Postgres runs only explicitly. This way `cargo test` is not affected by the speed of running docker containers and executing Postgres queries.